File: ios/Barem/Barem/Engine/CalcHistoryRepository.swift
import Foundation

final class CalcHistoryRepository {
    static let shared = CalcHistoryRepository()
    static let maxEntries = 30

    private let fileURL: URL

    private init() {
        let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        fileURL = dir.appendingPathComponent("calc_history.json")
    }

    func list() -> [CalcHistoryRecord] {
        guard let data = try? Data(contentsOf: fileURL) else { return [] }
        return (try? JSONDecoder().decode([CalcHistoryRecord].self, from: data)) ?? []
    }

    func delete(id: String) {
        save(list().filter { $0.id != id })
    }

    func saveFromSession(form: MemurFormState, result: SalaryResult) {
        let record = CalcHistoryRecord(
            id: UUID().uuidString,
            savedAtEpochMs: Int64(Date().timeIntervalSince1970 * 1000),
            form: form,
            kadroSummary: CalcSession.kadroSummaryLine(form: form),
            netSummary: CalcSession.netSummaryLine(result: result),
            periodLabel: result.periodLabel
        )
        var next = [record] + list()
        var seen = Set<String>()
        next = next.filter { seen.insert(dedupeKey($0.form)).inserted }
        save(Array(next.prefix(Self.maxEntries)))
    }

    private func dedupeKey(_ form: MemurFormState) -> String {
        [
            form.hizmetSinifi, form.unvan, form.kadroDetay,
            form.hesapYili.map(String.init), form.derece.map(String.init),
            form.medeniHal, form.kademe.map(String.init), form.kidemYili.map(String.init),
            form.il, form.ilce, form.topluSozlesme,
            String(form.cocukUst6), String(form.cocukAlt6),
        ].map { $0 ?? "" }.joined(separator: "|")
    }

    private func save(_ records: [CalcHistoryRecord]) {
        guard let data = try? JSONEncoder().encode(records) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: app/src/main/java/tr/erdaldemir/barem/data/CalcHistoryRepository.kt
package tr.erdaldemir.barem.data

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import tr.erdaldemir.barem.domain.model.MemurFormState
import tr.erdaldemir.barem.domain.model.SalaryResult
import tr.erdaldemir.barem.domain.session.CalcHistoryRecord
import tr.erdaldemir.barem.domain.session.CalcSessionStore
import java.io.File
import java.util.UUID

class CalcHistoryRepository(context: Context) {

    private val appContext = context.applicationContext
    private val gson = Gson()
    private val file: File by lazy { File(appContext.filesDir, "calc_history.json") }

    fun list(): List<CalcHistoryRecord> = loadAll().sortedByDescending { it.savedAtEpochMs }

    fun find(id: String): CalcHistoryRecord? = loadAll().find { it.id == id }

    fun delete(id: String) {
        saveAll(loadAll().filterNot { it.id == id })
    }

    fun saveFromSession(form: MemurFormState, result: SalaryResult) {
        val record = CalcHistoryRecord(
            id = UUID.randomUUID().toString(),
            savedAtEpochMs = System.currentTimeMillis(),
            form = form,
            kadroSummary = CalcSessionStore.kadroSummaryLine(form).orEmpty(),
            netSummary = CalcSessionStore.netSummaryLine(result).orEmpty(),
            periodLabel = result.periodLabel,
        )
        val next = (listOf(record) + loadAll())
            .distinctBy { historyDedupeKey(it.form) }
            .take(MAX_ENTRIES)
        saveAll(next)
    }

    private fun historyDedupeKey(form: MemurFormState): String = buildString {
        append(form.hizmetKolu).append('|')
        append(form.hizmetSinifi).append('|')
        append(form.meslekKod).append('|')
        append(form.unvan).append('|')
        append(form.kadroDetay).append('|')
        append(form.hesapYili).append('|')
        append(form.derece).append('|')
        append(form.medeniHal).append('|')
        append(form.kademe).append('|')
        append(form.kidemYili).append('|')
        append(form.il).append('|')
        append(form.ilce).append('|')
        append(form.topluSozlesme).append('|')
        append(form.cocukUst6).append('|')
        append(form.cocukAlt6)
    }

    private fun loadAll(): List<CalcHistoryRecord> {
        if (!file.exists()) return emptyList()
        return runCatching {
            val type = object : TypeToken<List<CalcHistoryRecord>>() {}.type
            gson.fromJson<List<CalcHistoryRecord>>(file.readText(), type) ?: emptyList()
        }.getOrDefault(emptyList())
    }

    private fun saveAll(records: List<CalcHistoryRecord>) {
        file.writeText(gson.toJson(records))
    }

    companion object {
        const val MAX_ENTRIES = 30
    }
}
